### PR TITLE
feat: add OpinionatedVestingWalletFactory for 4-year vest / 1-year cliff

### LINKED ISSUE #6: Create opinionated factory
Given our intended usage of the vesting contract, we should create an opinionated version of the factory that hard-codes the duration/cliff at 4/1 years respectively. It can inherit the existing factory and override the `create` function.

### LINKED ISSUE #8: Index the owner and beneficiary in the factory creation event
Currently, the `VestingWalletCreated` event indexes only the `walletAddress`, but it would be very useful to be able to search by `ownerAddress` and `beneficiaryAddress` as well.

### PR DESCRIPTION
**Pull request by rabmarut**
Closes #6.
Closes #8.

After #5 is merged, I will change the base branch for this PR to `main`.

### CODE REVIEW

File: src/OpinionatedVestingWalletFactory.sol [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        address ownerAddress,
+        address beneficiaryAddress,
+        uint64 startTimestamp
+    ) public virtual returns (address) {
+        return _create(
+            ownerAddress,
+            beneficiaryAddress,
+            startTimestamp,
+            4 * 52 weeks, // `years` and `months` are deprecated
+            52 weeks
```
Comment (gerrrg) [inline src/OpinionatedVestingWalletFactory.sol:26]:
```suggestion
            4 * 365 days, // `years` and `months` are deprecated
            365 days
```

Comment (gerrrg) [reply src/OpinionatedVestingWalletFactory.sol:26]: 52 * 7 = 364 🤷 
If you want to get really nit-picky and talk about leap years, we could make the duration 4 * 365 days + 1 days.

File: src/OpinionatedVestingWalletFactory.sol
Comment (gerrrg) [inline src/OpinionatedVestingWalletFactory.sol:1]: I acknowledge the general usefulness of this opinionated vesting wallet, but assume you realize that it is not terribly useful for (75% of) our own use, correct?

Comment (gerrrg) [reply src/OpinionatedVestingWalletFactory.sol:1]: Also I don't love the name, but I'm not sure what a better one would be. 

`VestingWallet4YearVest1YearCliff`?
`VestingWallet4yV1yC`?
`4YearVestWallet1YearCliff`?
`VestingWallet4YearVest1YearCliffWithClawback`?
`VestingWallet4YearVest1YearCliffWithClawbackAndGuardedRelease`?

File: src/VestingWalletWithCliffAndClawbackFactory.sol
Comment (gerrrg) [inline src/VestingWalletWithCliffAndClawbackFactory.sol:1]: I get what you're trying to do, but this contract seems pretty useless. Is there a compelling reason not to just have `OpinionatedFactory` point to (the old version of) `VestingWalletWithCliffAndClawbackFactory`?

File: test/OpinionatedVestingWalletFactory.t.sol [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+import "../src/OpinionatedVestingWalletFactory.sol";
+import "../src/VestingWalletWithCliffAndClawback.sol";
+
+contract OpinionatedVestingWalletFactoryTest is Test {
+    OpinionatedVestingWalletFactory public factory;
+    VestingWalletWithCliffAndClawback public wallet;
+
+    address owner = vm.addr(0x1);
+    address recipient = vm.addr(0x2);
```
Comment (gerrrg) [inline test/OpinionatedVestingWalletFactory.t.sol:14]:
```suggestion
    address beneficiary = vm.addr(0x2);
```

File: test/OpinionatedVestingWalletFactory.t.sol [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    function setUp() public {
+        factory = new OpinionatedVestingWalletFactory();
+        address walletAddress = factory.create(owner, recipient, uint64(block.timestamp));
```
Comment (gerrrg) [inline test/OpinionatedVestingWalletFactory.t.sol:18]:
```suggestion
        address walletAddress = factory.create(owner, beneficiary, uint64(block.timestamp));
```

File: test/OpinionatedVestingWalletFactory.t.sol [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        wallet = VestingWalletWithCliffAndClawback(payable(walletAddress));
+    }
+
+    function testWalletIsFromFactory() view public {
+        assert(factory.isWalletFromFactory(address(wallet)));
+    }
+
+    function testVestDuration() view public {
+        assert(wallet.duration() == 4 * 52 weeks);
```
Comment (gerrrg) [inline test/OpinionatedVestingWalletFactory.t.sol:27]:
```suggestion
        assert(wallet.duration() == 4 * 365 days);
```

File: test/OpinionatedVestingWalletFactory.t.sol [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    function testCliffDuration() view public {
+        assert(wallet.cliffDuration() == 52 weeks);
```
Comment (gerrrg) [inline test/OpinionatedVestingWalletFactory.t.sol:31]:
```suggestion
        assert(wallet.cliffDuration() == 365 days);
```

**Review by gerrrg — COMMENTED** [review]
I don't have terribly strong feelings about most of the comments in this review, but in short:

* Base factory feels like excessive abstraction
* I think we should workshop the `OpinionatedFactory` name
* Nit-pick: 52 weeks < 1 year
* minor: `recipient` -> `beneficiary`

**Comment by rabmarut** [comment]
Closing as @gerrrg correctly pointed out that we don't really need this factory.

**Pull request closed without merging**